feat: globalMiddleware option

felix-dolderer opened the pull request:
### 🔗 Linked issue

Closes #114

### ❓ Type of change

- [x] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

Adds a configuration option that adds a global middleware. This option requires being logged in for all pages (unless otherwise specified on the individual page)

Exposes and auto-imports the middleware functions (`hankoLoggedIn`, `hankoLoggedOut`) so they can be imported in a simple and typesafe way.

### ℹ️ Help / Review wanted

1. I'm not sure if auto-importing the existing middleware functions (`hankoLoggedIn`, `hankoLoggedOut`) is necessary or the best way to make them accessible for building custom middleware on top of them.
2. How can I write tests for the logged-in behaviour? I logged in with my demo Hanko-API, took my current cookie and used that to make sure my tests were passing. This approach is obviously hard to replicate for other people who want to run the tests. (_Addition_: And it breaks automated testing.)

autofix-troubleshooter commented:
Hi! I'm the <a href="https://autofix.ci"><img alt="autofix logo" src="https://autofix.ci/logo/logo.svg" width="16"><strong>autofix.ci</strong></a> troubleshooter bot.

It looks like you correctly set up a CI job that uses the autofix.ci GitHub Action, but the [**autofix.ci GitHub App**](https://github.com/apps/autofix-ci/) has not been installed for this repository. This means that autofix.ci unfortunately does not have the permissions to fix this pull request. If you are the repository owner, please install the app and then restart the CI workflow! 😃

danielroe reviewed (commented):
sorry I missed this PR! one comment/question

felix-dolderer commented on README.md:
```diff
@@ -63,17 +64,27 @@ Check out the [Hanko documentation](https://docs.hanko.io/guides/vue) to learn m
 
 - `hanko-logged-in` will prevent access to a page unless you are logged in. (It will redirect you to `redirects.login` instead, and then redirect back to this page once you login. You can disable this behaviour by setting `redirects.followRedirect` to `false`.)
 - `hanko-logged-out` will prevent access to a page unless you are logged out. (It will redirect you to `redirects.success` when you log in, and otherwise to `redirects.home`.)
+- `hanko-allow-all` will allow all users access to page, even if the `globalMiddleware` option is set to `true`. (If the `globalMiddleware` option is not set to `true`, this middleware has no effect.)
```
Oh yes, I like that idea. Not sure when I'll get to it though.

felix-dolderer replied:
@danielroe I spent some time on this yesterday. And I really can't figure out how to define the types for a custom route meta.

I tried this:```

declare module 'nuxt/dist/pages/composables' {
  interface PageMeta {
    hanko?: {
      allow: 'all' | 'logged-in' | 'logged-out'
      deny: 'logged-in' | 'logged-out'
    }
  }
}
```

I also tried to declare it in some other modules (each time running dev:prepare) and never got it to be picked up by TypeScript so I get type-hints.

felix-dolderer commented:
I noticed none of the auth modules for nuxt have tests for the logged in state. I decided to fix that and added tests that mock useHanko and useRequestEvent in a way that the middleware accepts the request as authenticated.

felix-dolderer committed:
feat: add global middleware support and enhance route protection

- Introduced `globalMiddleware` option in the Nuxt configuration to protect all pages by default.
- Added new route middleware for handling access based on user authentication status.
- Updated README to include details on global middleware usage and configuration.
- Created example pages demonstrating the use of `allow` and `deny` settings in pageMeta.
- Enhanced test coverage for global middleware scenarios, including client-side and SSR tests.

felix-dolderer committed:
fix html warnings

felix-dolderer committed:
fix type import

felix-dolderer committed:
feat: add type definitions for page metadata and enhance type preparation

- Introduced a new `page-meta.d.ts` file to define the `hanko` property in the `PageMeta` interface.
- Updated the module to prepare type references for the new page metadata definitions.

felix-dolderer committed:
feat: implement conditional navigation based on global middleware setting

- Added a computed property to determine if global middleware is active.
- Updated the navigation structure in `app.vue` to conditionally render links based on the global middleware state.
- Enhanced the module to include `globalMiddleware` in the Nuxt configuration for improved route protection.

felix-dolderer committed:
refactor: move HankoMeta type definitions to page-meta-global.d.ts

- Replaced the old page-meta.d.ts file with a new page-meta-global.d.ts to avoid automatic inclusion in the TypeScript project.
- Updated type preparation in the module to reference the new file for conditional loading based on globalMiddleware settings.
- Removed the outdated page-meta.d.ts file.

felix-dolderer committed:
feat: enhance navigation with test environment support

- Added a computed property to check if the application is running in a test environment.
- Updated the navigation structure in `app.vue` to conditionally render links based on the test environment and global middleware state.

felix-dolderer committed:
stabilize tests navigation display

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

MickL commented:
Is anything still blocking this PR? Personally I think it is crucial to have a global middleware instead of setting it per page.

IMO it would be fine if Hanko would just export its middleware, see: https://github.com/nuxt-modules/hanko/issues/114#issuecomment-4266378097

felix-dolderer commented:
> Is anything still blocking this PR? Personally I think it is crucial to have a global middleware instead of setting it per page.
> 
> IMO it would be fine if Hanko would just export its middleware, see: [#114 (comment)](https://github.com/nuxt-modules/hanko/issues/114#issuecomment-4266378097)

This PR also exposes the existing middleswares so you can build your own pattern around that.

I'm not aware of anything blocking this, it just seems to be low priority and I'm not a maintainer so I cannot just merge the PR.
